src/Headless/Avalonia.Headless.NUnit/AvaloniaTestMethodCommand.cs

```diff
@@ -45,22 +50,75 @@ private AvaloniaTestMethodCommand(
 
     public static TestCommand ProcessCommand(HeadlessUnitTestSession session, TestCommand command)
     {
-        return ProcessCommand(session, command, new List<Action>(), new List<Action>());
+        return ProcessCommand(session, command, [], []);
     }
 
-    private static TestCommand ProcessCommand(HeadlessUnitTestSession session, TestCommand command, List<Action> before, List<Action> after)
+    private static TestCommand ProcessCommand(HeadlessUnitTestSession session, TestCommand command, List<Action> before, List<Action<TestExecutionContext>> after)
     {
         if (command is BeforeAndAfterTestCommand beforeAndAfterTestCommand)
         {
             if (s_beforeTest.GetValue(beforeAndAfterTestCommand) is Action<TestExecutionContext> beforeTest)
             {
-                Action<TestExecutionContext> beforeAction = c => before.Add(() => beforeTest(c));
-                s_beforeTest.SetValue(beforeAndAfterTestCommand, beforeAction);
+                var setUpTearDownInfo = beforeTest.Target?.GetType()
+                    .GetField("setUpTearDown", BindingFlags.Instance | BindingFlags.Public);
+                if (setUpTearDownInfo != null && setUpTearDownInfo.GetValue(beforeTest.Target) is SetUpTearDownItem setUpTearDown
+                    && s_setUpMethods.GetValue(setUpTearDown) is List<IMethodInfo> setUpMethods)
+                {
+                    Action<TestExecutionContext> beforeAction = c =>
+                    {
+                        before.AddRange(setUpMethods.Select<IMethodInfo, Action>(m => async void () =>
+                        {
+                            var result = m.Invoke(c.TestObject, []);
+                            if (result is Task task)
```

Comment:
The logic to await `Task/ValueTask` is now repeated 3 times in this class. Extracting it to an helper method would help keeping the code clean.